src/lib-compression/istream-zstd.c

```diff
@@ -0,0 +1,237 @@
+/* Copyright (c) 2013-2018 Dovecot authors, see the included COPYING file */
+
+#include "lib.h"
+
+#ifdef HAVE_ZSTD
+
+#include "buffer.h"
+#include "istream-private.h"
+#include "istream-zlib.h"
+#include <zstd.h>
+
+struct zstd_istream {
+	struct istream_private istream;
+	ZSTD_DStream *dstream;
+
+	struct stat last_parent_statbuf;
+	ZSTD_inBuffer input;
+	size_t input_true_size;
+	ZSTD_outBuffer output;
+	size_t next_read;
+	ssize_t buffer_size;
+	bool marked : 1;
+	bool log_errors : 1;
+};
+
+static void i_stream_zstd_close(struct iostream_private *stream,
+				bool close_parent)
+{
+	struct zstd_istream *zstream = (struct zstd_istream *)stream;
+
+	ZSTD_freeDStream(zstream->dstream);
+	zstream->dstream = NULL;
+
+	i_free(zstream->input.src);
+	i_free(zstream->output.dst);
+	if (close_parent)
+		i_stream_close(zstream->istream.parent);
+}
+
+static size_t i_stream_zstd_decompress(struct zstd_istream *zstream)
+{
+	struct istream_private *stream = (struct istream_private *)zstream;
+	const unsigned char *data;
+	size_t size;
+
+	if (zstream->next_read || !stream->istream.eof) {
+		// move everything to back of buffer
+		memmove(zstream->input.src,
+			(void *)((long)zstream->input.src + zstream->input.pos),
```

Comment:
use CONST_PTR_OFFSET. Also you can use PTR_OFFSET(zstream->input.src, 0) to non-constify it.